File: Cpp/Algorithm/src/Chapter3/reverseKGroup_25.cpp
#include "chapter3.h"
ListNode* myAlgo::ReverseKGroup_25::reverseKGroup(ListNode* head, int k)
{
	ListNode* dummyHead = new ListNode(0);;
	dummyHead->next = head;
	ListNode* prev = dummyHead;
	ListNode* kStart = head;
	ListNode* kEnd = head;
	ListNode* nxtKstart = head;
	while (kEnd)
	{
		for (size_t i = 0; i < k - 1 && kEnd != NULL; i++)
		{
			kEnd = kEnd->next;
		}
		if (kEnd == NULL)
		{
			prev->next = kStart;
			break;
		}
		nxtKstart = kEnd->next;
		kEnd->next = NULL;
		reverse(kStart);
		prev->next = kEnd;
		prev = kStart;
		kStart = nxtKstart;
		kEnd = kStart;
	}
	delete dummyHead;
	return dummyHead->next;
}

ListNode* myAlgo::ReverseKGroup_25::reverse(ListNode* head)
{
	if (head == NULL || head->next == NULL)
	{
		return head;
	}
	ListNode* newHead = reverse(head->next);
	head->next->next = head;
	head->next = NULL;
	return newHead;
}

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/FourthRound/Chapter03/reverseKGroup_25.java
package algo.tzashinorpu.FourthRound.Chapter03;

public class reverseKGroup_25 {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummyHead = new ListNode(0);
        dummyHead.next = head;
        ListNode prev = dummyHead;
        ListNode start = head;
        ListNode nextStart;
        ListNode end = head;
        while (end != null) {
            int tmpK = k;
            while (end != null && (--tmpK) > 0) {
                end = end.next;
            }
            if (end == null) {
                break;
            }
            nextStart = end.next;
            end.next = null;
            prev.next = reverse(start);
            start.next = nextStart;
            prev = start;
            start = nextStart;
            end = start;
            prev.next = start;
        }
        return dummyHead.next;
    }

    /*private ListNode reverse(ListNode start) {
        ListNode dummyHead = new ListNode(0);
        dummyHead.next = start;
        ListNode prev = dummyHead;
        ListNode curr = start;
        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }*/
    private ListNode reverse(ListNode start) {
        if (start == null || start.next == null) {
            return start;
        }
        ListNode newHead = reverse(start.next);
        start.next.next = start;
        start.next = null;
        return newHead;
    }
}
